File: src/utils/color.ts
// ASS colour helpers.
//
// ASS stores colours as "&HAABBGGRR" — alpha first, then blue/green/red (reversed
// from the usual RGB). Alpha 00 = fully opaque, FF = fully transparent.
// The UI uses HTML "#RRGGBB" pickers, so we convert and preserve alpha separately.

/** "&H00FFCC00" -> "#00CCFF" (drops alpha). Falls back to white. */
export function assColorToHex(ass: string): string {
  const m = ass.match(/&H([0-9A-Fa-f]{2})?([0-9A-Fa-f]{6})/);
  if (!m) return "#FFFFFF";
  const bgr = m[2];
  const bb = bgr.slice(0, 2);
  const gg = bgr.slice(2, 4);
  const rr = bgr.slice(4, 6);
  return `#${rr}${gg}${bb}`.toUpperCase();
}

/** Extract the alpha byte ("00" when absent). */
export function assAlpha(ass: string): string {
  const m = ass.match(/&H([0-9A-Fa-f]{2})[0-9A-Fa-f]{6}/);
  return m ? m[1].toUpperCase() : "00";
}

/** "#00CCFF" + alpha -> "&HAABBGGRR". */
export function hexToAssColor(hex: string, alpha = "00"): string {
  const h = hex.replace("#", "");
  const rr = h.slice(0, 2);
  const gg = h.slice(2, 4);
  const bb = h.slice(4, 6);
  return `&H${alpha}${bb}${gg}${rr}`.toUpperCase();
}

File: native/Sources/GlyphlineCore/Core/Color.swift
// ASS colour helpers (ported from ../../src/utils/color.ts).
//
// ASS stores colours as "&HAABBGGRR" — alpha first, then blue/green/red (reversed
// from RGB). Alpha 00 = opaque, FF = transparent. The UI uses "#RRGGBB", so we
// convert and keep alpha separately.

import Foundation

/// "&H00FFCC00" -> "#00CCFF" (drops alpha). Falls back to white.
public func assColorToHex(_ ass: String) -> String {
    guard let g = firstMatchGroups(#"&H([0-9A-Fa-f]{2})?([0-9A-Fa-f]{6})"#, ass), let bgr = g[2] else {
        return "#FFFFFF"
    }
    let bb = bgr.prefix(2)
    let gg = bgr.dropFirst(2).prefix(2)
    let rr = bgr.dropFirst(4).prefix(2)
    return "#\(rr)\(gg)\(bb)".uppercased()
}

/// Extract the alpha byte ("00" when absent).
public func assAlpha(_ ass: String) -> String {
    guard let g = firstMatchGroups(#"&H([0-9A-Fa-f]{2})[0-9A-Fa-f]{6}"#, ass), let a = g[1] else {
        return "00"
    }
    return a.uppercased()
}

/// "#00CCFF" + alpha -> "&HAABBGGRR".
public func hexToAssColor(_ hex: String, alpha: String = "00") -> String {
    let h = hex.replacingOccurrences(of: "#", with: "")
    let rr = h.prefix(2)
    let gg = h.dropFirst(2).prefix(2)
    let bb = h.dropFirst(4).prefix(2)
    return "&H\(alpha)\(bb)\(gg)\(rr)".uppercased()
}

// First match's capture groups (index 0 = whole match), nil for non-participating.
func firstMatchGroups(_ pattern: String, _ input: String) -> [String?]? {
    guard let re = try? NSRegularExpression(pattern: pattern) else { return nil }
    let ns = input as NSString
    guard let m = re.firstMatch(in: input, range: NSRange(location: 0, length: ns.length)) else { return nil }
    return (0..<m.numberOfRanges).map { i in
        let r = m.range(at: i)
        return r.location == NSNotFound ? nil : ns.substring(with: r)
    }
}
